src/components/Form.tsx

```diff
@@ -0,0 +1,272 @@
+				if (handlers.onBuild) handlers.onBuild(args[1]);
+				break;
+			case 'onFocus':
+				if (handlers.onFocus) handlers.onFocus(args[1]);
+				break;
+			case 'onBlur':
+				if (handlers.onBlur) handlers.onBlur(args[1]);
+				break;
+			case 'onInitialized':
+				if (handlers.onInitialized) handlers.onInitialized();
+				break;
+			default:
+				break;
+		}
+	}
+};
+
+const createWebformInstance = async (
+	FormConstructor: FormConstructor | undefined,
+	formSource: FormSource,
+	element: MutableRefObject<HTMLDivElement>['current'],
+	options: FormProps['options'] = {},
+) => {
+	if (!options?.events) {
+		options.events = getDefaultEmitter();
+	}
+	const promise = FormConstructor
+		? new FormConstructor(element, formSource, options)
+		: new FormClass(element, formSource, options);
+	const instance = await promise.ready;
+	return instance;
+};
+
+// Define effective props (aka I want to rename these props but also maintain backwards compatibility)
+const getEffectiveProps = (props: FormProps) => {
+	const {
+		Formio: FormioProp,
+		formioform,
+		form,
+		src,
+		formReady,
+		onFormReady,
+	} = props;
+	const formConstructor = FormioProp !== undefined ? FormioProp : formioform;
+
+	const formSource = form !== undefined ? form : src;
+
+	const formReadyCallback =
+		onFormReady !== undefined ? onFormReady : formReady;
+
+	return { formConstructor, formSource, formReadyCallback };
+};
+
+const Form = (props: FormProps) => {
+	const [formInstance, setFormInstance] = useState<Webform | null>(null);
+	const renderElement = useRef<HTMLDivElement | null>(null);
+	const prevFormDefinition = useRef<JSON | null>(null);
+
+	const { formConstructor, formSource, formReadyCallback } =
+		getEffectiveProps(props);
+	const {
+		src,
+		form,
+		submission,
+		url,
+		options,
+		formioform,
+		formReady,
+		Formio,
+		...handlers
+	} = props;
+
+	useEffect(() => () => formInstance?.destroy(), [formInstance]);
+
+	useEffect(() => {
+		if (renderElement.current === null) {
+			console.warn('Form element not found');
+			return;
+		}
+		if (typeof formSource === 'string') {
+			createWebformInstance(
+				formConstructor,
+				formSource,
+				renderElement.current,
+				options,
+			).then((instance) => {
+				if (instance) {
+					instance.src = formSource;
+					if (formReadyCallback) formReadyCallback(instance);
+					setFormInstance(instance);
+				}
+			});
+		} else if (
+			typeof formSource === 'object' &&
+			!isEqual(formSource, prevFormDefinition.current)
+		) {
+			prevFormDefinition.current = cloneDeep(formSource);
+			createWebformInstance(
+				formConstructor,
+				formSource,
+				renderElement.current,
+				options,
+			).then((instance) => {
+				if (instance && formSource) {
+					instance.form = formSource;
+					if (url) {
+						instance.url = url;
+					}
+					if (formReadyCallback) formReadyCallback(instance);
+					setFormInstance(instance);
+				}
+			});
+		}
+	}, [formConstructor, formSource, url, formReadyCallback, options]);
```

Comment:
```
useEffect(() => {
    const createInstance = async () => {
        
        if (renderElement.current === null) {
            console.warn('Form element not found');
            return;
        }

        if (typeof formSource === 'object') {
            // If formSource is an object and hasn't changed, return early
            if (isEqual(formSource, prevFormDefinition.current)) {
                return;
            }
            
            prevFormDefinition.current = cloneDeep(formSource);
        }

        const instance = await createWebformInstance(formConstructor, formSource, renderElement.current, options);

        if (instance) {
            
            if (typeof formSource === 'string') {
                instance.src = formSource;
            } else if (typeof formSource === 'object') {
                instance.form = formSource;
                
                if (url) {
                    instance.url = url;
                }
            }
            
            if (formReadyCallback) {
                formReadyCallback(instance);
            }
            
            setFormInstance(instance);
        }
    };

    createInstance();

}, [formConstructor, formSource, url, formReadyCallback, options]);

```

@brendanbond pls check if it looks more readable.

Reply:
@gyanendrasinghpanwar yeah, I toyed with something like this originally, but for whatever reason I really wanted to move the `createInstance` method out of the render function scope. I do like this a little better though for the lack of duplication - I'll make the change

Reply:
Updated!